internal/cri/server/list_pod_sandbox_metrics_linux.go

```diff
@@ -0,0 +1,1038 @@
+				})
+			}
+
+			// Add swap metrics if available
+			if s.Memory.Swap != nil {
+				metrics = append(metrics, &runtime.Metric{
+					Name:        "container_memory_swap",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.Swap.Usage},
+				})
+			}
+
+			// Add usage failcnt if available
+			if s.Memory.Usage != nil {
+				metrics = append(metrics, &runtime.Metric{
+					Name:        "container_memory_failcnt",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_COUNTER,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.Usage.Failcnt},
+				})
+			}
+
+		}
+
+	case *cg2.Metrics:
+		if s.Memory != nil {
+			metrics = append(metrics, []*runtime.Metric{
+				{
+					Name:        "container_memory_usage_bytes",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.Usage},
+				},
+				{
+					Name:        "container_memory_max_usage_bytes",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.UsageLimit},
+				},
+				{
+					Name:        "container_memory_working_set_bytes",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: getWorkingSetV2(s.Memory)},
+				},
+				{
+					Name:        "container_memory_rss",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.Anon},
+				},
+				{
+					Name:        "container_memory_cache",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.File},
+				},
+				{
+					Name:        "container_memory_kernel_usage",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.KernelStack},
+				},
+				{
+					Name:        "container_memory_mapped_file",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.FileMapped},
+				},
+				{
+					Name:        "container_memory_swap",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_GAUGE,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: s.Memory.SwapUsage},
+				},
+				{
+					Name:        "container_memory_failcnt",
+					Timestamp:   timestamp,
+					MetricType:  runtime.MetricType_COUNTER,
+					LabelValues: labels,
+					Value:       &runtime.UInt64Value{Value: 0}, // cgroups v2 doesn't expose failcnt, provide 0
```

Comment:
memory.events.local oom maybe? or is this a part of the TODO below on hierarchical